x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/SearchableSnapshotDirectory.java

```diff
@@ -506,10 +506,13 @@ private void prewarmCache(ActionListener<Void> listener, Supplier<Boolean> cance
 
                     final AtomicLong prefetchedBytes = new AtomicLong(0L);
                     try (var fileListener = new RefCountingListener(ActionListener.runBefore(completionListener.acquire().map(v -> {
-                        if (prefetchedBytes.get() == 0L) {
+                        // we can't report a file as being partially recovered from disk and partially from the blob store, but this is
+                        // something that can happen for fully mounted searchable snapshots. So we make a choice here: if less than 50%
+                        // of the file is prewarmed from the blob store then it is marked as reused from cache.
+                        if ((prefetchedBytes.get() / (double) file.length()) < 0.5d) {
```

Comment:
Is there a chance this could break for a very small segment files?
I am thinking about example where 1/3 parts were in disk already and we only downloaded 2/3.

I wonder if we should consider changins `boolean reused` to `long reused` in `FileDetail` and consider a file fully recovered when `recovered < length && reused + recovered == length`?

Reply:
I'm not sure to understand why it would break for small files? We're comparing the number of bytes prewarmed with the total file length so the number of parts do not really matter. Also, most small files are likely to be either fully prewarmed or fully reused.

> I wonder if we should consider changins `boolean reused` to `long reused` in `FileDetail` and consider a file fully recovered when `recovered < length && reused + recovered == length`?

We could but this is a larger change that I don't think is really worth doing since it only makes sense for cold cache.

Reply:
I am trying to understand if there is something that guarantee that the following would not happen:
* 2/3 are already cached locally and would not be downloaded
* 1/3 is downloaded updating a stats so that it expects something would complete download before reporting it as finished.